src/java/org/apache/cassandra/locator/HttpServiceConnector.java

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.locator;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.apache.cassandra.config.DurationSpec;
+import org.apache.cassandra.exceptions.ConfigurationException;
+
+import static java.lang.String.format;
+
+public class HttpServiceConnector
+{
+    protected final String metadataServiceUrl;
+    protected final int requestTimeoutMs;
+
+    public HttpServiceConnector(String serviceUrl, int requestTimeoutMs)
+    {
+        this.metadataServiceUrl = serviceUrl;
+        this.requestTimeoutMs = requestTimeoutMs;
+    }
+
+    public final String apiCall(String query) throws IOException
+    {
+        return apiCall(metadataServiceUrl, query, "GET", ImmutableMap.of(), 200);
+    }
+
+    public final String apiCall(String query, Map<String, String> extraHeaders) throws IOException
+    {
+        return apiCall(metadataServiceUrl, query, "GET", extraHeaders, 200);
+    }
+
+    public String apiCall(String url,
+                          String query,
+                          String method,
+                          Map<String, String> extraHeaders,
+                          int expectedResponseCode) throws IOException
+    {
+        HttpURLConnection conn = null;
+        try
+        {
+            conn = (HttpURLConnection) new URL(url + query).openConnection();
+            extraHeaders.forEach(conn::setRequestProperty);
+            conn.setRequestMethod(method);
+            conn.setConnectTimeout(requestTimeoutMs);
+            if (conn.getResponseCode() != expectedResponseCode)
+                throw new HttpException(conn.getResponseCode(), conn.getResponseMessage());
+
+            List<String> response = new ArrayList<>();
+
+            try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(conn.getInputStream())))
+            {
+                String line;
+                while ((line = bufferedReader.readLine()) != null) {
+                    response.add(line);
+                }
+            }
+
+            return response.stream().collect(Collectors.joining(System.lineSeparator()));
+        }
+        finally
+        {
+            if (conn != null)
+                conn.disconnect();
+        }
+    }
+
+    public static String resolveMetadataUrl(Properties properties, String keyProperty)
+    {
+        String parsedUrl = properties.getProperty(keyProperty);
+
+        try
+        {
+            if (parsedUrl == null)
+                throw new IllegalStateException(String.format("%s was not specified", keyProperty));
```

Comment:
I guess it should be the `ConfigurationException` type, right? As it might not be handled properly when the node starts